Dataset View Component

dannon opened the pull request:
Want to iterate on this and fix any bugs before the release.  Some preview displays can take a moment, so I need to add a loading indicator there.  Also needs tweaks to header handling in some of the subcomponents.  Includes unit tests, but will need selenium updates and additions.

I'm still working out the details of making it possible to have 'preferred' display applications for particular datatypes.  These will likely be both admin and user specified -- like, when someone wants to view an h5, we don't display the binary view, we use hd5web by default instead.  These will, of course, be opinionated so maybe we should focus on binary or otherwise not-useful-primary-display types first.  Second PR coming.

https://github.com/user-attachments/assets/005cd37d-71f6-427a-b9b3-188e1b6700ff

This needs several tests updated but could use feedback on the basic functionality.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

dannon committed:
Sketch out initial dataset view component

dannon committed:
Add view to itemurls type, incremental refactoring

dannon committed:
fix preview iframe sizing

dannon committed:
Refactor dataset preview tab styling and remove unnecessary CSS rules

dannon committed:
Refine dataset component attribute display

dannon committed:
Fix linting post-rebase

dannon committed:
Adjust routing for dataset view

dannon committed:
streamline tab management and improve error handling

dannon committed:
Consolidate dataset view routes

dannon committed:
Typing, cleanup, formatting, etc. in DatasetView

dannon committed:
Add testing for datasetview component

dannon committed:
Fix typing for datasetview test?

dannon committed:
Dump ts for testing datasetview; too much aggravation to work around for little benefit to actually *testing*

dannon committed:
Cleanup in DatasetActions

dannon committed:
Handle rerun; cleanup in contentOptions

dannon committed:
Fix showError and onError functions in DatasetActions component

dannon committed:
More styling of the header, include state and collapsible toggle

dannon committed:
Datset component iframe loading indicator fixes

Clearer indication of datset preview

dannon committed:
Adjust headline to match card font-weights -- bold title, normal hid

dannon committed:
Minor animation for dataset header in dataset component

dannon committed:
Refactor header state management to use persistent toggle

guerler commented on client/src/components/History/Content/ContentOptions.vue:
```diff
@@ -66,11 +61,16 @@ function onDisplay($event: MouseEvent) {
+function onRerun() {
+    // Emit the rerun event to be handled by parent component
+    emit("rerun");
+}
```
You can call `$emit('rerun')` in the template and remove this function.

guerler commented on client/src/components/Dataset/DatasetView.vue:
```diff
@@ -0,0 +1,373 @@
+
+.state-running,
+.state-upload {
+    background-color: #cff4fc;
```
We should only use theme colors, hard-coded colors like this break theming.

guerler commented on client/src/components/Dataset/DatasetView.vue:
```diff
@@ -0,0 +1,373 @@
+                <BTab title="Preview" class="iframe-card">
+                    <div class="preview-container position-relative h-100">
+                        <!-- Loading indicator for iframe -->
+                        <div v-if="iframeLoading" class="iframe-loading">
```
Can we use the LoadingSpan here instead of custom implementations? It ensures consistency across the user interface and promotes reusability.

guerler commented on client/src/components/Dataset/DatasetView.vue:
```diff
@@ -0,0 +1,373 @@
+    flex-direction: column;
+    padding: 0;
+    overflow: hidden;
+}
```
Customization is fine, but we should introduce a reusable Tabs component so this can be reused going forward.

guerler replied:
This is not so critical right now that hardcoded colors have been replaced with theme-colors. Thx.

dannon committed:
Fix scrolling in dataset component non-iframe tabs -- preview override got dropped somewhere

dannon committed:
swap to standard loading span

guerler commented on client/src/components/Dataset/DatasetView.vue:
```diff
@@ -0,0 +1,377 @@
+    edit: TABS.EDIT,
+    error: TABS.ERROR,
+};
+
```
Instead of maintaining both `tabIndexToName` and `tabNameToIndex`, we can simplify by using a single array:
`const tabs = ["preview", "details", "visualize", "edit", "error"]`.
Since tabNameToIndex is only used once, using `tabs.indexOf(name)` is performant and avoids unnecessary pre-indexing. This improves readability and maintainability.

dannon replied:
I think the performance argument doesn't really hold water, but I see pros and cons to both approaches. I personally prefer keeping logic out of templates - it makes the code easier to follow and maintains a cleaner separation of concerns. Sure, our current approach with these mappings is a bit wordier, but when I look at template code with stuff like TABS.indexOf(activeTab) and arrow functions in event handlers, it makes templates harder to scan quickly. Templates should be about structure and appearance, not logic. The extra mapping code is a small price to pay for keeping that separation clean and making the relationships between tabs more explicit. I'd rather have slightly more code that's clearer about what it's doing.

In any event, here's what that would look like and I can add it if we feel very strongly about this.
```diff
diff --git a/client/src/components/Dataset/DatasetView.vue b/client/src/components/Dataset/DatasetView.vue
index 5979d6b3e9c..1cc5edcc29d 100644
--- a/client/src/components/Dataset/DatasetView.vue
+++ b/client/src/components/Dataset/DatasetView.vue
@@ -38,35 +38,10 @@ const props = defineProps({
     },
 });
 
-type TabNames = "preview" | "details" | "visualize" | "edit" | "error";
-
-const TABS = {
-    PREVIEW: 0,
-    DETAILS: 1,
-    VISUALIZE: 2,
-    EDIT: 3,
-    ERROR: 4,
-} as const;
-
-type TabIndex = (typeof TABS)[keyof typeof TABS];
-
-const tabIndexToName: Record<number, TabNames> = {
-    [TABS.PREVIEW]: "preview",
-    [TABS.DETAILS]: "details",
-    [TABS.VISUALIZE]: "visualize",
-    [TABS.EDIT]: "edit",
-    [TABS.ERROR]: "error",
-};
-
-const tabNameToIndex: Record<TabNames, number> = {
-    preview: TABS.PREVIEW,
-    details: TABS.DETAILS,
-    visualize: TABS.VISUALIZE,
-    edit: TABS.EDIT,
-    error: TABS.ERROR,
-};
-
-const activeTab = ref<TabIndex>(TABS.PREVIEW);
+const TABS = ["preview", "details", "visualize", "edit", "error"] as const;
+type TabName = typeof TABS[number];
+
+const activeTab = ref<TabName>(TABS[0]);
 
 const dataset = computed(() => datasetStore.getDataset(props.datasetId));
 const isLoading = computed(() => datasetStore.isLoadingDataset(props.datasetId));
@@ -89,25 +64,23 @@ function toggleHeader() {
     toggleHeaderCollapse();
 }
 
-function onTabChange(tabIndex: number) {
-    if (!(tabIndex in tabIndexToName)) {
-        console.error("Invalid tab index received:", tabIndex);
+function onTabChange(tabName: TabName) {
+    if (!TABS.includes(tabName)) {
+        console.error("Invalid tab name received:", tabName);
         return;
     }
 
-    const tabName = tabIndexToName[tabIndex];
-
-    if (tabIndex === TABS.ERROR && !showError.value) {
+    if (tabName === "error" && !showError.value) {
         return;
     }
 
     // Reset iframe loading state when switching to Preview tab
-    if (tabIndex === TABS.PREVIEW) {
+    if (tabName === "preview") {
         iframeLoading.value = true;
     }
 
     const basePath = `/datasets/${props.datasetId}`;
-    const targetPath = tabIndex === TABS.PREVIEW ? basePath : `${basePath}/${tabName}`;
+    const targetPath = tabName === "preview" ? basePath : `${basePath}/${tabName}`;
 
     // Always push the navigation request; vue-router handles duplicates if the path is identical.
     router.push(targetPath);
@@ -116,26 +89,31 @@ function onTabChange(tabIndex: number) {
 // Set the active tab based on the current route
 function setActiveTabFromProp() {
     const currentTabName = props.tab as string; // Use the prop value
-    let targetIndex = tabNameToIndex[currentTabName as TabNames] as TabIndex;
 
-    if (targetIndex === undefined) {
-        targetIndex = TABS.PREVIEW;
+    // Check if tab name is valid
+    if (!TABS.includes(currentTabName as TabName)) {
         if (currentTabName !== "preview") {
             const previewPath = `/datasets/${props.datasetId}`;
             router.replace(previewPath);
             return;
         }
-    } else if (targetIndex === TABS.ERROR && !showError.value) {
-        targetIndex = TABS.PREVIEW;
+        activeTab.value = "preview";
+        return;
+    }
+
+    // Handle error tab special case
+    if (currentTabName === "error" && !showError.value) {
         if (currentTabName === "error") {
             const previewPath = `/datasets/${props.datasetId}`;
             router.replace(previewPath);
             return;
         }
+        activeTab.value = "preview";
+        return;
     }
 
-    if (activeTab.value !== targetIndex) {
-        activeTab.value = targetIndex;
+    if (activeTab.value !== currentTabName) {
+        activeTab.value = currentTabName as TabName;
     }
 }
 
@@ -146,7 +124,7 @@ watch(() => props.tab, setActiveTabFromProp, { immediate: true });
 watch(
     () => props.datasetId,
     () => {
-        if (activeTab.value === TABS.PREVIEW) {
+        if (activeTab.value === "preview") {
             iframeLoading.value = true;
         }
     }
@@ -178,7 +156,7 @@ watch(
                 </span>
                 <span v-if="dataset.genome_build" class="dbkey">
                     <span v-localize class="prompt">database</span>
-                    <BLink class="value font-weight-bold" data-label="Database/Build" @click="onTabChange(TABS.EDIT)">{{
+                    <BLink class="value font-weight-bold" data-label="Database/Build" @click="onTabChange('edit')">{{
                         dataset.genome_build
                     }}</BLink>
                 </span>
@@ -191,7 +169,7 @@ watch(
         <!-- Tab container - make it grow to fill remaining space and handle overflow -->
         <div class="dataset-tabs-container flex-grow-1 overflow-hidden">
             <!-- Make BTabs fill its container and use flex column layout internally -->
-            <BTabs pills card lazy class="h-100 d-flex flex-column" :value="activeTab" @input="onTabChange">
+            <BTabs pills card lazy class="h-100 d-flex flex-column" :active-tab-index="TABS.indexOf(activeTab)" @input="tabIndex => onTabChange(TABS[tabIndex])">
                 <BTab title="Preview" class="iframe-card">
                     <div class="preview-container position-relative h-100">
                         <!-- Loading indicator for iframe -->

```

guerler replied:
Yes, this looks much better to me. I see your point, but I think minimal logic like `TABS.indexOf(activeTab)` in the templates is fine, if it bothers too much, keeping the `const TABS` would be fine imo (then you could still do `activeTab.value === TABS.PREVIEW`), but I don't think that we need to create a cross-referenced data structure for tab handling. Sorry for the bike shedding!

dannon replied:
I really think there's a benefit to keeping as little logic in templates as possible. Sure, it's easy enough to follow in a smallish component with only a few of these instances, but this is actually related to the other point above about emits — it's much cleaner and easier to track what's happening when all template logic is bound locally and handled in script, whether it's an emit or something else.

And, I appreciate the review and don't mind a bit of bike-shedding, but with the imminent freeze, I think we might benefit from just getting this done and following up on non-functional refactorings separately.

I still need to update the tests here, unless someone else wants to play with that.

guerler replied:
Thanks!

dannon committed:
Suggested approach in review for tab management

dannon committed:
Keep enum for tabs

dannon committed:
Fix active tab binding logic.

dannon committed:
Overhaul color application; use theme vars where possible.

dannon committed:
Refactor state styles to use predefined vars.

guerler reviewed (approved):
This is a useful feature. Some remaining issues may need to be revisited, like e.g. moving the dataset info to the `Dataset Information` panel, but we can address those during the freeze period. Thank you for adding it.

dannon committed:
Fixed header animation flickering during tab changes by replacing CSS

animations with Vue transitions for smoother state management. This
ensures the header details animation only occurs when intentionally
toggling the header, not during tab navigation.

dannon committed:
Improve dataset header collapse animation to create a smoother, more
immediate transition effect. Replace transform-based animations with
comprehensive transitions for all properties including max-height,
margins, and padding for a fluid collapsing behavior.

dannon committed:
DatasetView -- swap order of viz and details

dannon committed:
Relocate rerun to details.

mvdbeek commented on client/src/components/Dataset/DatasetView.vue:
```diff
@@ -0,0 +1,384 @@
+
+const dataset = computed(() => datasetStore.getDataset(props.datasetId));
+const isLoading = computed(() => datasetStore.isLoadingDataset(props.datasetId));
+const displayUrl = computed(() => `/datasets/${props.datasetId}/display/?preview=true`);
```
This is different in #20154, but there we do need 
```suggestion
const displayUrl = computed(() => withPrefix(`/datasets/${props.datasetId}/display/?preview=true`));
```

mvdbeek commented on client/src/components/Dataset/DatasetView.vue:
```diff
@@ -0,0 +1,384 @@
+import { computed, ref, watch } from "vue";
+import { useRouter } from "vue-router/composables";
+
+import { STATES } from "@/components/History/Content/model/states";
```
```suggestion
import { withPrefix } from "@/utils/redirect";
import { STATES } from "@/components/History/Content/model/states";
```
